subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -3912,9 +3915,15 @@ static int setup_cert(struct net_app_ctx *app_ctx, void *cert)
 int lwm2m_engine_start(struct lwm2m_ctx *client_ctx,
 		       char *peer_str, u16_t peer_port)
 {
+	struct net_if *iface = net_if_get_default();
```

Comment:
At some point (no need to do this for this PR), we should convert this to select correct network interface depending on destination IP address. This is very much needed if there are multiple network interfaces in the system. One such example is ethernet + VLAN support which will create multiple network interfaces in the system. For IPv4, we already have `net_if_ipv4_select_src_iface()` function, but corresponding IPv6 function is missing (I will add it soon). Also more generic version like `net_if_select_src_iface(struct sockaddr *addr)` might be useful (I will add that too).

Reply:
@jukkar This PR can wait till those functions are merged, and I'll rework to use them.

Reply:
Ok, the PR is #9322, I will add you as a reviewer to that.

Reply:
@jukkar I tested your PR and unfortunately it doesn't work well as well as I had hoped.

The network interface needs to be "up" prior to the net_app_init_udp_client() call in lwm2m_engine_start().  There are times when we perform network access such as DNS lookup inside net_app_init_udp_client().  Of course, if we need to use DNS then I don't know if the result will be an IPv4 or IPv6 address for the destination.  And I need that destination to choose which interface to make sure is "up"... :/

Reply:
>  And I need that destination to choose which interface to make sure is "up"... :/

Which shows again that working with "interfaces" in the code which is otherwise intended to work with "addresses and ports" is wrong...

Reply:
@mike-scott Ok, we can go with the current solution if you really need to have this functionality in lwm2m code. Eventually it would be better to remove this iface check from individual component and let system to send event when everything is ready to be connected. But this might take some time before we can have this functionality. I can start to investigate this but it will probably go to 1.14 as merge window is closing soonish.